src/runtime_src/core/tools/xbutil2/OO_AieClockFreq.cpp

```diff
@@ -0,0 +1,187 @@
+#include "core/common/query_requests.h"
+#include "OO_AieClockFreq.h"
+#include "tools/common/XBUtilities.h"
+
+namespace XBU = XBUtilities;
+
+// 3rd Party Library - Include Files
+#include <boost/algorithm/string/join.hpp>
+#include <boost/format.hpp>
+#include <boost/program_options.hpp>
+
+namespace po = boost::program_options;
+namespace qr = xrt_core::query;
+
+// System - Include Files
+#include <iostream>
+
+#define to_mega(x) x/1000000
+
+// ----- H E L P E R M E T H O D S ------------------------------------------
+static double get_aie_part_freq(const std::shared_ptr<xrt_core::device>& device, uint32_t part_id)
+{
+  double freq = 0;
+  try {
+    freq = static_cast<double>(to_mega(xrt_core::device_query<qr::aie_get_freq>(device, part_id)));
+  }
+  catch (const std::exception &e) {
+    std::cerr << boost::format("ERROR: Failed to read clock frequency of AIE partition(%d)\n %s\n") % part_id % e.what();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  return freq;
+}
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+OO_AieClockFreq::OO_AieClockFreq( const std::string &_longName, bool _isHidden )
+    : OptionOptions(_longName, _isHidden, "AIE clock frequency operations" )
+    , m_device("")
+    , m_partition_id(1)
+    , m_get(false)
+    , m_freq("")
+    , m_help(false)
+{
+  m_optionsDescription.add_options()
+    ("device,d", po::value<decltype(m_device)>(&m_device)->multitoken(), "The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest")
+    ("partition,p", po::value<decltype(m_partition_id)>(&m_partition_id), "The Partition id of AIE")
+    ("set,s", po::value<decltype(m_freq)>(&m_freq), "Frequency value (Hz) to set given AIE partition to (eg: 100K, 312.5M, 5G)")
+    ("get,g", po::bool_switch(&m_get), "Read the frequency of given AIE partition")
+    ("help,h", po::bool_switch(&m_help), "Help to use this sub-command")
+  ;
+}
+
+void
+OO_AieClockFreq::execute(const SubCmdOptions& _options) const
+{
+
+  XBU::verbose("SubCommand option: AIE Clock");
+
+  XBU::verbose("Option(s):");
+  for (auto & aString : _options)
+    XBU::verbose(std::string(" ") + aString);
+
+  // Honor help option first
+  if (std::find(_options.begin(), _options.end(), "--help") != _options.end()) {
+    printHelp();
+    return;
+  }
+
+  // Parse sub-command ...
+  po::variables_map vm;
+
+  try {
+    po::store(po::command_line_parser(_options).options(m_optionsDescription).positional(m_positionalOptions).run(), vm);
+    po::notify(vm); // Can throw
+  }
+  catch (po::error& e) {
+    std::cerr << "ERROR: " << e.what() << "\n\n";
+    printHelp();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  // Exit if neither action or device specified
+  if(m_help || m_device.empty()) {
+    printHelp();
+    return;
+  }
+
+  // Check if set/get is used
+  if(!m_get && m_freq.length() == 0) {
+    std::cerr << "ERROR: Neither `set` nor `get' is used" << std::endl;
```

Comment:
The message could be a little bit more terse.  Something like:
`ERROR: Missing either a 'set' or 'get' option.`
Then print the help.